### ISSUE TITLE
createBroadcast implementation causes performance issues

### ISSUE THREAD

**Issue by IIIristraM** [styled-components/styled-components#931]
this lines
```
return () => {
    listeners = listeners.filter(item => item !== listener)
 }
```
produce a lot of new objects which leads to a frequent garbage collections, which take sometimes up to 250 ms.

this can be probably avoided by changing `listeners` from array to a simple hash object like {  [id]: listner }, and the lines above to

```
return () => {
    delete listeners[id]
 }
```

**Comment by mxstbr** [comment]
Great catch, PR welcome!

**Comment by philpl** [comment]
We should probably just switch to iamstarkov/theming soon.

But it depends on `brcast` which uses the same filtering logic unfortunately: https://github.com/vesparny/brcast/blob/master/index.js#L16

**Comment by vesparny** [comment]
I suspect what a bout using splice?

```js
listeners.splice(listeners.indexOf(listener), 1)
```

**Comment by vesparny** [comment]
I'll be happy to apply the patch to `brcast`.
Just wondering what's the best way to gain performance.

**Comment by mxstbr** [comment]
Wrote a very simple microbench (likely doing things wrong, please double check my code): https://esbench.com/bench/5950cbe599634800a034871c

On my machine the `delete listeners[key]` version is ~3x faster that `.splice` and ~20x faster than `.filter`!

**Comment by vesparny** [comment]
I have published a new version of `brcast` that addresses this issue.
Feel free to steal the code from there while you wait to switch to `theming`

**Comment by xcoderzach** [comment]
@schwers Can you make a pull request from that fork?  It would be very useful for that to be in the official release 😄

**Comment by vesparny** [comment]
Note I've released a new version of brcast fixing another issue.
It should also be a little bit more perdormant.

Do you want me to submit a PR that replaces SC implementation with brcast?

**Comment by xcoderzach** [comment]
@vesparny that would be awesome!  I'd use it immediately (I have a page that takes 2 seconds to unmount in production 😬 )

**Comment by vesparny** [comment]
Didn't you have plans for switching to theming?

**Comment by xcoderzach** [comment]
Looks like it fell by the wayside.  A pull request for that would be welcome too, but may be more difficult.

**Comment by mxstbr** [comment]
Yeah let's switch to theming, assuming all our existing tests pass, PR welcome!

**Comment by mxstbr** [comment]
This was fixed in #1048!
